tests/apollo-fetch.ts

```diff
@@ -0,0 +1,496 @@
+          && isEqual(variables, { personNum: 1 })) {
+        return simpleResult;
+      }
+
+      if (query === print(complexQueryWithTwoVars)
+          && isEqual(variables, { personNum: 1, filmNum: 1 })) {
+        return complexResult;
+      }
+      throw new Error('Invalid Query');
+    });
+    fetchMock.post(missingUrl, () => {
+      throw new Error('Network error');
+    });
+
+    fetchMock.post(unauthorizedUrl, 403);
+    fetchMock.post(serviceUnavailableUrl, 503);
+  });
+
+  afterEach(fetchMock.reset);
+
+  it('should not throw with no arguments', () => {
+    assert.doesNotThrow(createApolloFetch);
+  });
+
+  it('should call fetch', (done) => {
+    const fetcher = createApolloFetch();
+    const result = fetcher({query: print(sampleQuery)});
+    result.then((response) => {
+      assert.deepEqual(fetchMock.calls('/graphql').length, 1);
+      assert.deepEqual(response, JSON.parse(data));
+      done();
+    });
+  });
+
+  it('should call fetch with correct url', (done) => {
+    const uri = 'test';
+    const fetcher = createApolloFetch({uri});
+    const result = fetcher({query: print(sampleQuery)});
+    result.then((response) => {
+      //correct response
+      assert.deepEqual(response, JSON.parse(data));
+      //single call
+      assert.deepEqual(fetchMock.calls(uri).length, 1);
+
+      assert.deepEqual(fetchMock.lastCall(uri)[0], uri);
+      const options = fetchMock.lastCall(uri)[1];
+      const body = JSON.parse(options.body);
+      assert.deepEqual(options.method, 'POST');
+      assert.deepEqual(options.headers, {Accept: '*/*', 'Content-Type': 'application/json'});
+      assert.deepEqual(body.query, print(sampleQuery));
+      done();
+    });
+  });
+
+  it('should pass an error onto the Promise', () => {
+    const uri = 'error';
+    const fetcher = createApolloFetch({uri, customFetch: fetch});
+    const result = fetcher({query: print(sampleQuery)});
+    return assert.isRejected(result, mockError.throws, mockError.throws.message);
+  });
+
+// missingUrl
+
+  it('should catch on a network error', (done) => {
+    const fetcher = createApolloFetch({uri: unauthorizedUrl});
+    const result = fetcher({query: print(sampleQuery)});
+    result.then(expect.fail)
+    .catch((error) => {
+      assert.deepEqual(error.message, 'Network request failed with status 403 - \"Forbidden\"');
+      assert.isDefined(error.response);
+      assert.isDefined(error.parseError);
+      done();
+    });
+  });
+
+  it('should return a fail to parse response when fetch returns raw response', (done) => {
+    const fetcher = createApolloFetch({uri: '/raw'});
+    const result = fetcher({query: print(sampleQuery)});
+    result.then(expect.fail)
+    .catch((error) => {
+      assert.deepEqual(error.message, 'Network request failed to return valid JSON');
+      assert.isDefined(error.response);
+      assert.isDefined(error.raw);
+      assert.deepEqual(error.raw, unparsableData);
+      done();
```

Comment:
Btw, if you have a single promise, you can just return that instead of calling `done()` at the end every time. The test will pass if the promise resolves, and it will fail if the promise rejects.